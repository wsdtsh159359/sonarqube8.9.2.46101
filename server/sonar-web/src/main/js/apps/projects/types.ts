/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { ComponentQualifier } from '../../types/component';

export interface Project {
  analysisDate?: string;
  isFavorite?: boolean;
  key: string;
  leakPeriodDate?: string;
  measures: T.Dict<string>;
  name: string;
  projects?: number;
  qualifier: ComponentQualifier;
  tags: string[];
  visibility: T.Visibility;
  needIssueSync?: boolean;
}

export interface Facet {
  [value: string]: number;
}

export interface Facets {
  [property: string]: Facet;
}
