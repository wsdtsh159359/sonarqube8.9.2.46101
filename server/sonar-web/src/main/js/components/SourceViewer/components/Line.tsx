/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import { times } from 'lodash';
import * as React from 'react';
import { BranchLike } from '../../../types/branch-like';
import './Line.css';
import LineCode from './LineCode';
import LineCoverage from './LineCoverage';
import LineDuplicationBlock from './LineDuplicationBlock';
import LineIssuesIndicator from './LineIssuesIndicator';
import LineNumber from './LineNumber';
import LineSCM from './LineSCM';

interface Props {
  branchLike: BranchLike | undefined;
  displayAllIssues?: boolean;
  displayCoverage: boolean;
  displayDuplications: boolean;
  displayIssueLocationsCount?: boolean;
  displayIssueLocationsLink?: boolean;
  displayIssues: boolean;
  displayLineNumberOptions?: boolean;
  displayLocationMarkers?: boolean;
  displaySCM?: boolean;
  duplications: number[];
  duplicationsCount: number;
  firstLineNumber: number;
  highlighted: boolean;
  highlightedLocationMessage: { index: number; text: string | undefined } | undefined;
  highlightedSymbols: string[] | undefined;
  issueLocations: T.LinearIssueLocation[];
  issuePopup: { issue: string; name: string } | undefined;
  issues: T.Issue[];
  last: boolean;
  line: T.SourceLine;
  loadDuplications: (line: T.SourceLine) => void;
  onIssueChange: (issue: T.Issue) => void;
  onIssuePopupToggle: (issueKey: string, popupName: string, open?: boolean) => void;
  onIssuesClose: (line: T.SourceLine) => void;
  onIssueSelect: (issueKey: string) => void;
  onIssuesOpen: (line: T.SourceLine) => void;
  onIssueUnselect: () => void;
  onLocationSelect: ((x: number) => void) | undefined;
  onSymbolClick: (symbols: string[]) => void;
  openIssues: boolean;
  previousLine: T.SourceLine | undefined;
  renderDuplicationPopup: (index: number, line: number) => React.ReactNode;
  scroll?: (element: HTMLElement) => void;
  scrollToUncoveredLine?: boolean;
  secondaryIssueLocations: T.LinearIssueLocation[];
  selectedIssue: string | undefined;
  verticalBuffer?: number;
}

const LINE_HEIGHT = 18;

export default class Line extends React.PureComponent<Props> {
  handleIssuesIndicatorClick = () => {
    if (this.props.openIssues) {
      this.props.onIssuesClose(this.props.line);
      this.props.onIssueUnselect();
    } else {
      this.props.onIssuesOpen(this.props.line);

      const { issues } = this.props;
      if (issues.length > 0) {
        this.props.onIssueSelect(issues[0].key);
      }
    }
  };

  render() {
    const {
      branchLike,
      displayAllIssues,
      displayCoverage,
      displayDuplications,
      displayIssueLocationsCount,
      displayIssueLocationsLink,
      displayLineNumberOptions,
      displayLocationMarkers,
      highlightedLocationMessage,
      displayIssues,
      displaySCM = true,
      duplications,
      duplicationsCount,
      firstLineNumber,
      highlighted,
      highlightedSymbols,
      issueLocations,
      issuePopup,
      issues,
      last,
      line,
      openIssues,
      previousLine,
      scrollToUncoveredLine,
      secondaryIssueLocations,
      selectedIssue,
      verticalBuffer
    } = this.props;

    const className = classNames('source-line', {
      'source-line-highlighted': highlighted,
      'source-line-filtered': line.isNew,
      'source-line-filtered-dark':
        displayCoverage &&
        (line.coverageStatus === 'uncovered' || line.coverageStatus === 'partially-covered'),
      'source-line-last': last === true
    });

    const bottomPadding = verticalBuffer ? verticalBuffer * LINE_HEIGHT : undefined;

    // default is true
    const displayOptions = displayLineNumberOptions !== false;

    return (
      <tr className={className} data-line-number={line.line}>
        <LineNumber displayOptions={displayOptions} firstLineNumber={firstLineNumber} line={line} />

        {displaySCM && <LineSCM line={line} previousLine={previousLine} />}
        {displayIssues && !displayAllIssues ? (
          <LineIssuesIndicator
            issues={issues}
            issuesOpen={openIssues}
            line={line}
            onClick={this.handleIssuesIndicatorClick}
          />
        ) : (
          <td className="source-meta source-line-issues" />
        )}

        {displayDuplications && (
          <LineDuplicationBlock
            blocksLoaded={duplicationsCount > 0}
            duplicated={Boolean(line.duplicated)}
            index={0}
            key={0}
            line={this.props.line}
            onClick={this.props.loadDuplications}
            renderDuplicationPopup={this.props.renderDuplicationPopup}
          />
        )}

        {duplicationsCount > 1 &&
          times(duplicationsCount - 1, index => (
            <LineDuplicationBlock
              blocksLoaded={true}
              duplicated={duplications.includes(index + 1)}
              index={index + 1}
              key={index + 1}
              line={this.props.line}
              renderDuplicationPopup={this.props.renderDuplicationPopup}
            />
          ))}

        {displayCoverage && (
          <LineCoverage
            line={line}
            scroll={this.props.scroll}
            scrollToUncoveredLine={scrollToUncoveredLine}
          />
        )}

        <LineCode
          branchLike={branchLike}
          displayIssueLocationsCount={displayIssueLocationsCount}
          displayIssueLocationsLink={displayIssueLocationsLink}
          displayLocationMarkers={displayLocationMarkers}
          highlightedLocationMessage={highlightedLocationMessage}
          highlightedSymbols={highlightedSymbols}
          issueLocations={issueLocations}
          issuePopup={issuePopup}
          issues={issues}
          line={line}
          onIssueChange={this.props.onIssueChange}
          onIssuePopupToggle={this.props.onIssuePopupToggle}
          onIssueSelect={this.props.onIssueSelect}
          onLocationSelect={this.props.onLocationSelect}
          onSymbolClick={this.props.onSymbolClick}
          padding={bottomPadding}
          scroll={this.props.scroll}
          secondaryIssueLocations={secondaryIssueLocations}
          selectedIssue={selectedIssue}
          showIssues={openIssues || displayAllIssues}
        />
      </tr>
    );
  }
}
