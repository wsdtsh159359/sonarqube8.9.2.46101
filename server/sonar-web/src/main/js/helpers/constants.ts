/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { colors } from '../app/theme';
import { AlmKeys } from '../types/alm-settings';
import { ComponentQualifier } from '../types/component';
import { IssueScope, IssueType } from '../types/issues';

export const SEVERITIES = ['BLOCKER', 'CRITICAL', 'MAJOR', 'MINOR', 'INFO'];
export const STATUSES = ['OPEN', 'REOPENED', 'CONFIRMED', 'RESOLVED', 'CLOSED'];
export const ISSUE_TYPES: T.IssueType[] = [
  IssueType.Bug,
  IssueType.Vulnerability,
  IssueType.CodeSmell,
  IssueType.SecurityHotspot
];
export const SOURCE_SCOPES = [
  { scope: IssueScope.Main, qualifier: ComponentQualifier.File },
  { scope: IssueScope.Test, qualifier: ComponentQualifier.TestFile }
];
export const RULE_TYPES: T.RuleType[] = ['BUG', 'VULNERABILITY', 'CODE_SMELL', 'SECURITY_HOTSPOT'];
export const RULE_STATUSES = ['READY', 'BETA', 'DEPRECATED'];

export const CHART_COLORS_RANGE_PERCENT = [
  colors.green,
  colors.lightGreen,
  colors.yellow,
  colors.orange,
  colors.red
];

export const CHART_REVERSED_COLORS_RANGE_PERCENT = [
  colors.red,
  colors.orange,
  colors.yellow,
  colors.lightGreen,
  colors.green
];

export const RATING_COLORS = [
  colors.green,
  colors.lightGreen,
  colors.yellow,
  colors.orange,
  colors.red
];

export const PROJECT_KEY_MAX_LEN = 400;

export const ALM_DOCUMENTATION_PATHS = {
  [AlmKeys.Azure]: '/documentation/analysis/azuredevops-integration/',
  [AlmKeys.BitbucketServer]: '/documentation/analysis/bitbucket-integration/',
  [AlmKeys.BitbucketCloud]: '/documentation/analysis/bitbucket-cloud-integration/',
  [AlmKeys.GitHub]: '/documentation/analysis/github-integration/',
  [AlmKeys.GitLab]: '/documentation/analysis/gitlab-integration/'
};

export const IMPORT_COMPATIBLE_ALMS = [
  AlmKeys.Azure,
  AlmKeys.BitbucketServer,
  AlmKeys.GitHub,
  AlmKeys.GitLab
];
