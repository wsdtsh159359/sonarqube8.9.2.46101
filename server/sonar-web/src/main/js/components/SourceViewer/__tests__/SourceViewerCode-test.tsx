/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { shallow } from 'enzyme';
import * as React from 'react';
import { mockBranch } from '../../../helpers/mocks/branch-like';
import { mockIssue, mockSourceLine } from '../../../helpers/testMocks';
import { MetricKey } from '../../../types/metrics';
import Line from '../components/Line';
import SourceViewerCode from '../SourceViewerCode';

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot('default');
  expect(shallowRender({ issues: [mockIssue(false, { textRange: undefined })] })).toMatchSnapshot(
    'has file level issues'
  );
  expect(shallowRender({ hasSourcesAfter: true, hasSourcesBefore: true })).toMatchSnapshot(
    'has more sources'
  );
});

it('should correctly flag a line for scrolling', () => {
  const sources = [
    mockSourceLine({ line: 1, coverageStatus: 'covered', isNew: false }),
    mockSourceLine({ line: 2, coverageStatus: 'partially-covered', isNew: false }),
    mockSourceLine({ line: 3, coverageStatus: 'uncovered', isNew: true })
  ];
  let wrapper = shallowRender({ sources, metricKey: MetricKey.uncovered_lines });

  expect(
    wrapper
      .find(Line)
      .at(1)
      .props().scrollToUncoveredLine
  ).toBe(true);
  expect(
    wrapper
      .find(Line)
      .at(2)
      .props().scrollToUncoveredLine
  ).toBe(false);

  wrapper = shallowRender({
    sources,
    metricKey: MetricKey.new_uncovered_lines
  });

  expect(
    wrapper
      .find(Line)
      .at(1)
      .props().scrollToUncoveredLine
  ).toBe(false);
  expect(
    wrapper
      .find(Line)
      .at(2)
      .props().scrollToUncoveredLine
  ).toBe(true);
});

function shallowRender(props: Partial<SourceViewerCode['props']> = {}) {
  return shallow<SourceViewerCode>(
    <SourceViewerCode
      branchLike={mockBranch()}
      componentKey="foo"
      duplications={[]}
      duplicationsByLine={[]}
      hasSourcesAfter={false}
      hasSourcesBefore={false}
      highlightedLine={undefined}
      highlightedLocationMessage={undefined}
      highlightedLocations={undefined}
      highlightedSymbols={[]}
      issueLocationsByLine={{}}
      issuePopup={undefined}
      issues={[mockIssue(), mockIssue()]}
      issuesByLine={{}}
      loadDuplications={jest.fn()}
      loadingSourcesAfter={false}
      loadingSourcesBefore={false}
      loadSourcesAfter={jest.fn()}
      loadSourcesBefore={jest.fn()}
      onIssueChange={jest.fn()}
      onIssuePopupToggle={jest.fn()}
      onIssuesClose={jest.fn()}
      onIssueSelect={jest.fn()}
      onIssuesOpen={jest.fn()}
      onIssueUnselect={jest.fn()}
      onLocationSelect={jest.fn()}
      onSymbolClick={jest.fn()}
      openIssuesByLine={{}}
      renderDuplicationPopup={jest.fn()}
      selectedIssue={undefined}
      sources={[mockSourceLine(), mockSourceLine(), mockSourceLine()]}
      symbolsByLine={{}}
      {...props}
    />
  );
}
