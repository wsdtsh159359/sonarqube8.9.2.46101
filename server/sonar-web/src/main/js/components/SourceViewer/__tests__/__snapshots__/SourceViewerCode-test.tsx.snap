// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly: default 1`] = `
<div
  className="source-viewer-code"
>
  <table
    className="source-table"
  >
    <tbody>
      <Line
        branchLike={
          Object {
            "analysisDate": "2018-01-01",
            "excludedFromPurge": true,
            "isMain": false,
            "name": "branch-6.7",
          }
        }
        displayCoverage={true}
        displayDuplications={false}
        displayIssues={true}
        duplications={Array []}
        duplicationsCount={0}
        firstLineNumber={16}
        highlighted={false}
        issueLocations={Array []}
        issues={Array []}
        key="16"
        last={false}
        line={
          Object {
            "code": "<span class=\\"k\\">import</span> java.util.<span class=\\"sym-9 sym\\">ArrayList</span>;",
            "coverageStatus": "covered",
            "coveredConditions": 2,
            "duplicated": false,
            "isNew": true,
            "line": 16,
            "scmAuthor": "simon.brandhof@sonarsource.com",
            "scmDate": "2018-12-11T10:48:39+0100",
            "scmRevision": "80f564becc0c0a1c9abaa006eca83a4fd278c3f0",
          }
        }
        loadDuplications={[MockFunction]}
        onIssueChange={[MockFunction]}
        onIssuePopupToggle={[MockFunction]}
        onIssueSelect={[MockFunction]}
        onIssueUnselect={[MockFunction]}
        onIssuesClose={[MockFunction]}
        onIssuesOpen={[MockFunction]}
        onLocationSelect={[MockFunction]}
        onSymbolClick={[MockFunction]}
        openIssues={false}
        renderDuplicationPopup={[MockFunction]}
        scrollToUncoveredLine={false}
        secondaryIssueLocations={Array []}
      />
      <Line
        branchLike={
          Object {
            "analysisDate": "2018-01-01",
            "excludedFromPurge": true,
            "isMain": false,
            "name": "branch-6.7",
          }
        }
        displayCoverage={true}
        displayDuplications={false}
        displayIssues={true}
        duplications={Array []}
        duplicationsCount={0}
        firstLineNumber={16}
        highlighted={false}
        issueLocations={Array []}
        issues={Array []}
        key="16"
        last={false}
        line={
          Object {
            "code": "<span class=\\"k\\">import</span> java.util.<span class=\\"sym-9 sym\\">ArrayList</span>;",
            "coverageStatus": "covered",
            "coveredConditions": 2,
            "duplicated": false,
            "isNew": true,
            "line": 16,
            "scmAuthor": "simon.brandhof@sonarsource.com",
            "scmDate": "2018-12-11T10:48:39+0100",
            "scmRevision": "80f564becc0c0a1c9abaa006eca83a4fd278c3f0",
          }
        }
        loadDuplications={[MockFunction]}
        onIssueChange={[MockFunction]}
        onIssuePopupToggle={[MockFunction]}
        onIssueSelect={[MockFunction]}
        onIssueUnselect={[MockFunction]}
        onIssuesClose={[MockFunction]}
        onIssuesOpen={[MockFunction]}
        onLocationSelect={[MockFunction]}
        onSymbolClick={[MockFunction]}
        openIssues={false}
        previousLine={
          Object {
            "code": "<span class=\\"k\\">import</span> java.util.<span class=\\"sym-9 sym\\">ArrayList</span>;",
            "coverageStatus": "covered",
            "coveredConditions": 2,
            "duplicated": false,
            "isNew": true,
            "line": 16,
            "scmAuthor": "simon.brandhof@sonarsource.com",
            "scmDate": "2018-12-11T10:48:39+0100",
            "scmRevision": "80f564becc0c0a1c9abaa006eca83a4fd278c3f0",
          }
        }
        renderDuplicationPopup={[MockFunction]}
        scrollToUncoveredLine={false}
        secondaryIssueLocations={Array []}
      />
      <Line
        branchLike={
          Object {
            "analysisDate": "2018-01-01",
            "excludedFromPurge": true,
            "isMain": false,
            "name": "branch-6.7",
          }
        }
        displayCoverage={true}
        displayDuplications={false}
        displayIssues={true}
        duplications={Array []}
        duplicationsCount={0}
        firstLineNumber={16}
        highlighted={false}
        issueLocations={Array []}
        issues={Array []}
        key="16"
        last={true}
        line={
          Object {
            "code": "<span class=\\"k\\">import</span> java.util.<span class=\\"sym-9 sym\\">ArrayList</span>;",
            "coverageStatus": "covered",
            "coveredConditions": 2,
            "duplicated": false,
            "isNew": true,
            "line": 16,
            "scmAuthor": "simon.brandhof@sonarsource.com",
            "scmDate": "2018-12-11T10:48:39+0100",
            "scmRevision": "80f564becc0c0a1c9abaa006eca83a4fd278c3f0",
          }
        }
        loadDuplications={[MockFunction]}
        onIssueChange={[MockFunction]}
        onIssuePopupToggle={[MockFunction]}
        onIssueSelect={[MockFunction]}
        onIssueUnselect={[MockFunction]}
        onIssuesClose={[MockFunction]}
        onIssuesOpen={[MockFunction]}
        onLocationSelect={[MockFunction]}
        onSymbolClick={[MockFunction]}
        openIssues={false}
        previousLine={
          Object {
            "code": "<span class=\\"k\\">import</span> java.util.<span class=\\"sym-9 sym\\">ArrayList</span>;",
            "coverageStatus": "covered",
            "coveredConditions": 2,
            "duplicated": false,
            "isNew": true,
            "line": 16,
            "scmAuthor": "simon.brandhof@sonarsource.com",
            "scmDate": "2018-12-11T10:48:39+0100",
            "scmRevision": "80f564becc0c0a1c9abaa006eca83a4fd278c3f0",
          }
        }
        renderDuplicationPopup={[MockFunction]}
        scrollToUncoveredLine={false}
        secondaryIssueLocations={Array []}
      />
    </tbody>
  </table>
</div>
`;

exports[`should render correctly: has file level issues 1`] = `
<div
  className="source-viewer-code"
>
  <table
    className="source-table"
  >
    <tbody>
      <Line
        branchLike={
          Object {
            "analysisDate": "2018-01-01",
            "excludedFromPurge": true,
            "isMain": false,
            "name": "branch-6.7",
          }
        }
        displayCoverage={true}
        displayDuplications={false}
        displayIssues={true}
        duplications={Array []}
        duplicationsCount={0}
        firstLineNumber={16}
        highlighted={false}
        issueLocations={Array []}
        issues={Array []}
        key="0"
        last={false}
        line={
          Object {
            "code": "",
            "duplicated": false,
            "isNew": false,
            "line": 0,
          }
        }
        loadDuplications={[MockFunction]}
        onIssueChange={[MockFunction]}
        onIssuePopupToggle={[MockFunction]}
        onIssueSelect={[MockFunction]}
        onIssueUnselect={[MockFunction]}
        onIssuesClose={[MockFunction]}
        onIssuesOpen={[MockFunction]}
        onLocationSelect={[MockFunction]}
        onSymbolClick={[MockFunction]}
        openIssues={false}
        renderDuplicationPopup={[MockFunction]}
        scrollToUncoveredLine={false}
        secondaryIssueLocations={Array []}
      />
      <Line
        branchLike={
          Object {
            "analysisDate": "2018-01-01",
            "excludedFromPurge": true,
            "isMain": false,
            "name": "branch-6.7",
          }
        }
        displayCoverage={true}
        displayDuplications={false}
        displayIssues={true}
        duplications={Array []}
        duplicationsCount={0}
        firstLineNumber={16}
        highlighted={false}
        issueLocations={Array []}
        issues={Array []}
        key="16"
        last={false}
        line={
          Object {
            "code": "<span class=\\"k\\">import</span> java.util.<span class=\\"sym-9 sym\\">ArrayList</span>;",
            "coverageStatus": "covered",
            "coveredConditions": 2,
            "duplicated": false,
            "isNew": true,
            "line": 16,
            "scmAuthor": "simon.brandhof@sonarsource.com",
            "scmDate": "2018-12-11T10:48:39+0100",
            "scmRevision": "80f564becc0c0a1c9abaa006eca83a4fd278c3f0",
          }
        }
        loadDuplications={[MockFunction]}
        onIssueChange={[MockFunction]}
        onIssuePopupToggle={[MockFunction]}
        onIssueSelect={[MockFunction]}
        onIssueUnselect={[MockFunction]}
        onIssuesClose={[MockFunction]}
        onIssuesOpen={[MockFunction]}
        onLocationSelect={[MockFunction]}
        onSymbolClick={[MockFunction]}
        openIssues={false}
        renderDuplicationPopup={[MockFunction]}
        scrollToUncoveredLine={false}
        secondaryIssueLocations={Array []}
      />
      <Line
        branchLike={
          Object {
            "analysisDate": "2018-01-01",
            "excludedFromPurge": true,
            "isMain": false,
            "name": "branch-6.7",
          }
        }
        displayCoverage={true}
        displayDuplications={false}
        displayIssues={true}
        duplications={Array []}
        duplicationsCount={0}
        firstLineNumber={16}
        highlighted={false}
        issueLocations={Array []}
        issues={Array []}
        key="16"
        last={false}
        line={
          Object {
            "code": "<span class=\\"k\\">import</span> java.util.<span class=\\"sym-9 sym\\">ArrayList</span>;",
            "coverageStatus": "covered",
            "coveredConditions": 2,
            "duplicated": false,
            "isNew": true,
            "line": 16,
            "scmAuthor": "simon.brandhof@sonarsource.com",
            "scmDate": "2018-12-11T10:48:39+0100",
            "scmRevision": "80f564becc0c0a1c9abaa006eca83a4fd278c3f0",
          }
        }
        loadDuplications={[MockFunction]}
        onIssueChange={[MockFunction]}
        onIssuePopupToggle={[MockFunction]}
        onIssueSelect={[MockFunction]}
        onIssueUnselect={[MockFunction]}
        onIssuesClose={[MockFunction]}
        onIssuesOpen={[MockFunction]}
        onLocationSelect={[MockFunction]}
        onSymbolClick={[MockFunction]}
        openIssues={false}
        previousLine={
          Object {
            "code": "<span class=\\"k\\">import</span> java.util.<span class=\\"sym-9 sym\\">ArrayList</span>;",
            "coverageStatus": "covered",
            "coveredConditions": 2,
            "duplicated": false,
            "isNew": true,
            "line": 16,
            "scmAuthor": "simon.brandhof@sonarsource.com",
            "scmDate": "2018-12-11T10:48:39+0100",
            "scmRevision": "80f564becc0c0a1c9abaa006eca83a4fd278c3f0",
          }
        }
        renderDuplicationPopup={[MockFunction]}
        scrollToUncoveredLine={false}
        secondaryIssueLocations={Array []}
      />
      <Line
        branchLike={
          Object {
            "analysisDate": "2018-01-01",
            "excludedFromPurge": true,
            "isMain": false,
            "name": "branch-6.7",
          }
        }
        displayCoverage={true}
        displayDuplications={false}
        displayIssues={true}
        duplications={Array []}
        duplicationsCount={0}
        firstLineNumber={16}
        highlighted={false}
        issueLocations={Array []}
        issues={Array []}
        key="16"
        last={true}
        line={
          Object {
            "code": "<span class=\\"k\\">import</span> java.util.<span class=\\"sym-9 sym\\">ArrayList</span>;",
            "coverageStatus": "covered",
            "coveredConditions": 2,
            "duplicated": false,
            "isNew": true,
            "line": 16,
            "scmAuthor": "simon.brandhof@sonarsource.com",
            "scmDate": "2018-12-11T10:48:39+0100",
            "scmRevision": "80f564becc0c0a1c9abaa006eca83a4fd278c3f0",
          }
        }
        loadDuplications={[MockFunction]}
        onIssueChange={[MockFunction]}
        onIssuePopupToggle={[MockFunction]}
        onIssueSelect={[MockFunction]}
        onIssueUnselect={[MockFunction]}
        onIssuesClose={[MockFunction]}
        onIssuesOpen={[MockFunction]}
        onLocationSelect={[MockFunction]}
        onSymbolClick={[MockFunction]}
        openIssues={false}
        previousLine={
          Object {
            "code": "<span class=\\"k\\">import</span> java.util.<span class=\\"sym-9 sym\\">ArrayList</span>;",
            "coverageStatus": "covered",
            "coveredConditions": 2,
            "duplicated": false,
            "isNew": true,
            "line": 16,
            "scmAuthor": "simon.brandhof@sonarsource.com",
            "scmDate": "2018-12-11T10:48:39+0100",
            "scmRevision": "80f564becc0c0a1c9abaa006eca83a4fd278c3f0",
          }
        }
        renderDuplicationPopup={[MockFunction]}
        scrollToUncoveredLine={false}
        secondaryIssueLocations={Array []}
      />
    </tbody>
  </table>
</div>
`;

exports[`should render correctly: has more sources 1`] = `
<div
  className="source-viewer-code"
>
  <div
    className="source-viewer-more-code"
  >
    <Button
      className="js-component-viewer-source-before"
      onClick={[MockFunction]}
    >
      source_viewer.load_more_code
    </Button>
  </div>
  <table
    className="source-table"
  >
    <tbody>
      <Line
        branchLike={
          Object {
            "analysisDate": "2018-01-01",
            "excludedFromPurge": true,
            "isMain": false,
            "name": "branch-6.7",
          }
        }
        displayCoverage={true}
        displayDuplications={false}
        displayIssues={true}
        duplications={Array []}
        duplicationsCount={0}
        firstLineNumber={16}
        highlighted={false}
        issueLocations={Array []}
        issues={Array []}
        key="16"
        last={false}
        line={
          Object {
            "code": "<span class=\\"k\\">import</span> java.util.<span class=\\"sym-9 sym\\">ArrayList</span>;",
            "coverageStatus": "covered",
            "coveredConditions": 2,
            "duplicated": false,
            "isNew": true,
            "line": 16,
            "scmAuthor": "simon.brandhof@sonarsource.com",
            "scmDate": "2018-12-11T10:48:39+0100",
            "scmRevision": "80f564becc0c0a1c9abaa006eca83a4fd278c3f0",
          }
        }
        loadDuplications={[MockFunction]}
        onIssueChange={[MockFunction]}
        onIssuePopupToggle={[MockFunction]}
        onIssueSelect={[MockFunction]}
        onIssueUnselect={[MockFunction]}
        onIssuesClose={[MockFunction]}
        onIssuesOpen={[MockFunction]}
        onLocationSelect={[MockFunction]}
        onSymbolClick={[MockFunction]}
        openIssues={false}
        renderDuplicationPopup={[MockFunction]}
        scrollToUncoveredLine={false}
        secondaryIssueLocations={Array []}
      />
      <Line
        branchLike={
          Object {
            "analysisDate": "2018-01-01",
            "excludedFromPurge": true,
            "isMain": false,
            "name": "branch-6.7",
          }
        }
        displayCoverage={true}
        displayDuplications={false}
        displayIssues={true}
        duplications={Array []}
        duplicationsCount={0}
        firstLineNumber={16}
        highlighted={false}
        issueLocations={Array []}
        issues={Array []}
        key="16"
        last={false}
        line={
          Object {
            "code": "<span class=\\"k\\">import</span> java.util.<span class=\\"sym-9 sym\\">ArrayList</span>;",
            "coverageStatus": "covered",
            "coveredConditions": 2,
            "duplicated": false,
            "isNew": true,
            "line": 16,
            "scmAuthor": "simon.brandhof@sonarsource.com",
            "scmDate": "2018-12-11T10:48:39+0100",
            "scmRevision": "80f564becc0c0a1c9abaa006eca83a4fd278c3f0",
          }
        }
        loadDuplications={[MockFunction]}
        onIssueChange={[MockFunction]}
        onIssuePopupToggle={[MockFunction]}
        onIssueSelect={[MockFunction]}
        onIssueUnselect={[MockFunction]}
        onIssuesClose={[MockFunction]}
        onIssuesOpen={[MockFunction]}
        onLocationSelect={[MockFunction]}
        onSymbolClick={[MockFunction]}
        openIssues={false}
        previousLine={
          Object {
            "code": "<span class=\\"k\\">import</span> java.util.<span class=\\"sym-9 sym\\">ArrayList</span>;",
            "coverageStatus": "covered",
            "coveredConditions": 2,
            "duplicated": false,
            "isNew": true,
            "line": 16,
            "scmAuthor": "simon.brandhof@sonarsource.com",
            "scmDate": "2018-12-11T10:48:39+0100",
            "scmRevision": "80f564becc0c0a1c9abaa006eca83a4fd278c3f0",
          }
        }
        renderDuplicationPopup={[MockFunction]}
        scrollToUncoveredLine={false}
        secondaryIssueLocations={Array []}
      />
      <Line
        branchLike={
          Object {
            "analysisDate": "2018-01-01",
            "excludedFromPurge": true,
            "isMain": false,
            "name": "branch-6.7",
          }
        }
        displayCoverage={true}
        displayDuplications={false}
        displayIssues={true}
        duplications={Array []}
        duplicationsCount={0}
        firstLineNumber={16}
        highlighted={false}
        issueLocations={Array []}
        issues={Array []}
        key="16"
        last={false}
        line={
          Object {
            "code": "<span class=\\"k\\">import</span> java.util.<span class=\\"sym-9 sym\\">ArrayList</span>;",
            "coverageStatus": "covered",
            "coveredConditions": 2,
            "duplicated": false,
            "isNew": true,
            "line": 16,
            "scmAuthor": "simon.brandhof@sonarsource.com",
            "scmDate": "2018-12-11T10:48:39+0100",
            "scmRevision": "80f564becc0c0a1c9abaa006eca83a4fd278c3f0",
          }
        }
        loadDuplications={[MockFunction]}
        onIssueChange={[MockFunction]}
        onIssuePopupToggle={[MockFunction]}
        onIssueSelect={[MockFunction]}
        onIssueUnselect={[MockFunction]}
        onIssuesClose={[MockFunction]}
        onIssuesOpen={[MockFunction]}
        onLocationSelect={[MockFunction]}
        onSymbolClick={[MockFunction]}
        openIssues={false}
        previousLine={
          Object {
            "code": "<span class=\\"k\\">import</span> java.util.<span class=\\"sym-9 sym\\">ArrayList</span>;",
            "coverageStatus": "covered",
            "coveredConditions": 2,
            "duplicated": false,
            "isNew": true,
            "line": 16,
            "scmAuthor": "simon.brandhof@sonarsource.com",
            "scmDate": "2018-12-11T10:48:39+0100",
            "scmRevision": "80f564becc0c0a1c9abaa006eca83a4fd278c3f0",
          }
        }
        renderDuplicationPopup={[MockFunction]}
        scrollToUncoveredLine={false}
        secondaryIssueLocations={Array []}
      />
    </tbody>
  </table>
  <div
    className="source-viewer-more-code"
  >
    <Button
      className="js-component-viewer-source-after"
      onClick={[MockFunction]}
    >
      source_viewer.load_more_code
    </Button>
  </div>
</div>
`;
